### PR TITLE
account/usbwallet: abort usb enumeration after failures

### PR DESCRIPTION
**Pull request by holiman**
On my work machine (fedora 29 on Qubes), usb enumeration fails, and it would print out three error messages every second to tell me about it. 
This PR does not fix any underlying issue, but does abort USB enumeration after three failures (three per usb hub, so in total nine)
After this PR
```
INFO [06-05|10:54:43.822] Bumping default cache on mainnet         provided=1024 updated=4096
WARN [06-05|10:54:43.822] Sanitizing cache to Go's GC limits       provided=4096 updated=2590
INFO [06-05|10:54:43.827] Maximum peer count                       ETH=0 LES=0 total=0
INFO [06-05|10:54:43.828] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
ERROR[06-05|10:54:43.829] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=1 hub=ledger vendor=11415
ERROR[06-05|10:54:43.829] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=1 hub=trezor vendor=21324
ERROR[06-05|10:54:43.829] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=1 hub=trezor vendor=4617
ERROR[06-05|10:54:43.829] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=2 hub=ledger vendor=11415
ERROR[06-05|10:54:43.829] error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=2 hub=trezor vendor=21324
ERROR[06-05|10:54:43.829] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=2 hub=trezor vendor=4617
INFO [06-05|10:54:43.829] Starting peer-to-peer node               instance=Geth/v1.9.0-unstable-645756cd-20190604/linux-amd64/go1.11.2
INFO [06-05|10:54:43.829] Allocated trie memory caches             clean=647.00MiB dirty=647.00MiB
INFO [06-05|10:54:43.829] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=1.26GiB handles=2048
INFO [06-05|10:54:43.862] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient
INFO [06-05|10:54:43.862] Initialised chain configuration          config="{ChainID: 1 Homestead: 1150000 DAO: 1920000 DAOSupport: true EIP150: 2463000 EIP155: 2675000 EIP158: 2675000 Byzantium: 4370000 Constantinople: 7280000  ConstantinopleFix: 7280000 Engine: ethash}"
INFO [06-05|10:54:43.863] Disk storage enabled for ethash caches   dir=/home/user/.ethereum/geth/ethash count=3
INFO [06-05|10:54:43.863] Disk storage enabled for ethash DAGs     dir=/home/user/.ethash               count=2
INFO [06-05|10:54:43.863] Initialising Ethereum protocol           versions="[63 62]" network=1 dbversion=7
INFO [06-05|10:54:43.939] Loaded most recent local header          number=0 hash=d4e567…cb8fa3 td=17179869184 age=50y1mo3w
INFO [06-05|10:54:43.939] Loaded most recent local full block      number=0 hash=d4e567…cb8fa3 td=17179869184 age=50y1mo3w
INFO [06-05|10:54:43.939] Loaded most recent local fast block      number=0 hash=d4e567…cb8fa3 td=17179869184 age=50y1mo3w
INFO [06-05|10:54:43.939] Loaded local transaction journal         transactions=0 dropped=0
INFO [06-05|10:54:43.939] Regenerated local transaction journal    transactions=0 accounts=0
INFO [06-05|10:54:43.959] Allocated fast sync bloom                size=1.26GiB
INFO [06-05|10:54:43.993] New local node record                    seq=154 id=071a5f65ef12c6f1 ip=127.0.0.1 udp=0 tcp=30303
INFO [06-05|10:54:43.994] Started P2P networking                   self="enode://564c8a39cec34355b6e5d7603b912856e422c2770c7f7d7c4eb264fe0a30fb43b10fa8fc95103644b66efccc9bb5b6e793b5cf1ef1efce1dc3a9e1413dae791e@127.0.0.1:30303?discport=0"
INFO [06-05|10:54:44.015] IPC endpoint opened                      url=/home/user/.ethereum/geth.ipc
INFO [06-05|10:54:44.169] Initialized fast sync bloom              items=12356 errorrate=0.000 elapsed=209.076ms
ERROR[06-05|10:54:44.830] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=3 hub=trezor vendor=4617
ERROR[06-05|10:54:44.830] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=3 hub=ledger vendor=11415
ERROR[06-05|10:54:44.830] Error enumerating USB                    code="failed to initialize libusb: libusb: unknown error [code -99]" failcount=3 hub=trezor vendor=21324
^CINFO [06-05|10:55:13.009] Got interrupt, shutting down... 

```

### CODE REVIEW

File: accounts/usbwallet/hub.go [hunk 1]
```diff
@@ -156,11 +162,13 @@ func (hub *Hub) refreshWallets() {
 	}
 	infos, err := usb.Enumerate(hub.vendorID, 0)
 	if err != nil {
+		failcount := atomic.AddUint64(&(hub.enumerationFailures), 1)
 		if runtime.GOOS == "linux" {
 			// See rationale before the enumeration why this is needed and only on Linux.
 			hub.commsLock.Unlock()
 		}
-		log.Error("error enumerating USB enumeration: ", "code", err)
+		log.Error("Error enumerating USB", "code", err, "failcount", failcount,
```
Comment (karalabe) [inline accounts/usbwallet/hub.go]: Nits:

 * Please change the message to `Failed to enumerate USB devices`, we'll have an `err` field either way.
 * Please change the `code` to `err`, that's teh canonical way in the codebase.
 * Please move `err` to the end (the last field in a log can be arbitrarilly long, if you place the error as the first one, you'll end up with weird alignments).
 * Please move the hub and vendor to be the first and second fields respectively.

File: accounts/usbwallet/hub.go [hunk 2]
```diff
@@ -62,8 +63,9 @@ type Hub struct {
 	stateLock sync.RWMutex // Protects the internals of the hub from racey access
 
 	// TODO(karalabe): remove if hotplug lands on Windows
-	commsPend int        // Number of operations blocking enumeration
-	commsLock sync.Mutex // Lock protecting the pending counter and enumeration
+	commsPend           int        // Number of operations blocking enumeration
+	commsLock           sync.Mutex // Lock protecting the pending counter and enumeration
+	enumerationFailures uint64     // Number of times enumeration has failed
```
Comment (karalabe) [inline accounts/usbwallet/hub.go]: Could you rename this to `enumFails`? :D Looks nicer/neater.

**Comment by holiman** [comment]
fixed, ptal

File: accounts/usbwallet/hub.go [hunk 3]
```diff
@@ -64,6 +65,7 @@ type Hub struct {
 	// TODO(karalabe): remove if hotplug lands on Windows
 	commsPend int        // Number of operations blocking enumeration
 	commsLock sync.Mutex // Lock protecting the pending counter and enumeration
+	enumFails uint64     // Number of times enumeration has failed
```
Comment (karalabe) [inline accounts/usbwallet/hub.go]: Please use `uint32`:

> On ARM, x86-32, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. The first word in a variable or in an allocated struct, array, or slice can be relied upon to be 64-bit aligned.

https://golang.org/pkg/sync/atomic/#pkg-note-BUG

Comment (karalabe) [reply accounts/usbwallet/hub.go]: Sorry that I didn't catch this previously.

File: accounts/usbwallet/hub.go [hunk 4]
```diff
@@ -156,11 +162,13 @@ func (hub *Hub) refreshWallets() {
...
+		failcount := atomic.AddUint32(&(hub.enumFails), 1)
 		if runtime.GOOS == "linux" {
 			// See rationale before the enumeration why this is needed and only on Linux.
 			hub.commsLock.Unlock()
 		}
-		log.Error("error enumerating USB enumeration: ", "code", err)
+		log.Error("Failed to enumerate USB devices", "hub", hub.scheme,
+			"vendor", hub.vendorID, "failcount", failcount, "err", err)
 		return
 	}
```
Comment (karalabe) [inline accounts/usbwallet/hub.go]: Question: do we want to reset the counter if enumeration succeeds? That way occasional flukes don't disable the wallets.

Comment (karalabe) [reply accounts/usbwallet/hub.go]: Btw, you don't need paranthesis around `&(hub.enumFails)`

Comment (holiman) [reply accounts/usbwallet/hub.go]: No. I'll add that

**Pull request merged**